### DISCUSSION TITLE
a (temporary) alternative for safely using multiple BMS'es (with cell monitoring)

### DISCUSSION THREAD

**Discussion by arrow1800** [Louisvdw/dbus-serialbattery#276] [General]
hi Guys,

I just wanted to share a temporary alternative which i created for at least the time in which the serial bus driver is not able to support multiple bmsés. (but it maybe even has a purpose longer than that)

i used the beta version of the dbus driver so VRM and my Cerbogx at least can 'see' the different BMSes and also report their highest and lowest cell values (per pack) to the VEBUS and VRM.

then i used node-red (Venus os large image) to build a flow that **monitors each BMS**, extracts their max and minus values from the VEBUS and adjusts the grid-feed setting accordingly

a few reasons for this approach:

- i noticed my JK BMS from China is not capable of measuring the current in a appropriate way. Thus battery SOC is not correct
- as long as multiple bms'es aren't officially supported the current shown in the battery monitor is not correct as well. (shows only current from one bms)
- i also have a Victron smartshunt installed, which is calibrated way better then the chinese stuff. (its SOC calculaton can be trusted)

i am now able to do the following (adjusted every sec):

- when one of the cells gets to low, set gridfeed equal to outputpower of inverter and therefore battery becomes idle 
- when one of the cells gets to high, 
[flow.txt](https://github.com/Louisvdw/dbus-serialbattery/files/9878962/flow.txt)
set gridfeed to a negative value and therefore unload the batteries a bit till the high cell value is gone
- i am able to set the victron smartshunt as my primary battery monitor in the cerbo gx. 

![image](https://user-images.githubusercontent.com/12893401/198274380-ca17b7b2-5192-46a1-be04-5064cc35dcc1.png)

Love to hear your opinions!

**Comment by Dr-Gigavolt** [comment]
Hi and thanks for the idea. I just begun to play with the Node Red, I haven't hear about it before. I would like to create a virtual instance of the battery monitor and fill it with the sum of currents, average voltage, max. and min. of all cell voltages etc. and select this instance in the ESS. Is it in the Node Red possible?

**Reply by arrow1800** [reply]
hi, i just read this message. i just published a solution, take a look at this repo: https://github.com/arrow1800/Victron-VirtualBattery

**Reply by Dr-Gigavolt** [reply]
Hello and thank you. For me it was easier to code it in Python :-). I tried something with Node Red, but have no knowledge about Javascript. Now runs my Python program well, I've just published an update.

**Reply by mr-manuel** [reply]
Maybe this is something for you [venus-os_dbus-mqtt-battery](https://github.com/mr-manuel/venus-os_dbus-mqtt-battery)
With this driver you can feed a virtual battery over MQTT.
